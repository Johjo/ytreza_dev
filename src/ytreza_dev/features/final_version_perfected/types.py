from dataclasses import dataclass
from typing import Tuple


@dataclass(frozen=True, eq=True)
class Task:
    title: str
    url: str
    project_name: str


@dataclass(frozen=True, eq=True)
class NothingToDo:
    pass


@dataclass(frozen=True, eq=True)
class DoTheTask:
    task: Task


@dataclass(frozen=True, eq=True)
class ChooseTaskBetween:
    tasks: Tuple[Task, Task]


NextAction = NothingToDo | DoTheTask | ChooseTaskBetween

@dataclass(frozen=True, eq=True)
class Project:
    key: str
    name: str


@dataclass(frozen=True, eq=True)
class TaskBase:
    title: str
    url: str
    id: str

    def to_next(self) -> 'TaskNext':
        return TaskNext(title=self.title, url=self.url, id=self.id)

    def to_later(self) -> 'TaskLater':
        return TaskLater(title=self.title, url=self.url, id=self.id)

    def to_new(self) -> 'TaskNew':
        return TaskNew(title=self.title, url=self.url, id=self.id)

    def to_never(self) -> 'TaskNever':
        return TaskNever(title=self.title, url=self.url, id=self.id)


@dataclass(frozen=True, eq=True)
class TaskNew(TaskBase):
    pass


@dataclass(frozen=True, eq=True)
class TaskNext(TaskBase):
    pass


@dataclass(frozen=True, eq=True)
class TaskLater(TaskBase):
    pass

@dataclass(frozen=True, eq=True)
class TaskNever(TaskBase):
    pass


@dataclass(frozen=True, eq=True)
class ExternalProject:
    name: str
    key: str

@dataclass(frozen=True, eq=True)
class ExternalTask:
    name: str
    url: str
    id: str
    project: ExternalProject
