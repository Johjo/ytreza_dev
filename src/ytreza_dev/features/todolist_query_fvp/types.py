from dataclasses import dataclass
from typing import Tuple


@dataclass
class Task:
    title: str
    url: str


@dataclass
class NothingToDo:
    pass


@dataclass
class DoTheTask:
    task: Task


@dataclass
class ChooseTaskBetween:
    tasks: Tuple[Task, Task]


NextAction = NothingToDo | DoTheTask | ChooseTaskBetween


@dataclass()
class TaskBase:
    title: str
    url: str


@dataclass
class TaskNew(TaskBase):
    pass


@dataclass
class TaskNext(TaskBase):
    pass


@dataclass
class TaskLater(TaskBase):
    pass


