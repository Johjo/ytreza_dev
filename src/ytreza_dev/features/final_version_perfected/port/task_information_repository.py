import datetime
from abc import ABC, abstractmethod
from dataclasses import dataclass

from expression import Option

from ytreza_dev.features.final_version_perfected.types import Project


@dataclass
class TaskInformation:
    key: str
    url: str
    title: str
    project: Project
    due_date: Option[datetime.date]


class TaskInformationRepositoryPort(ABC):
    @abstractmethod
    def save(self, task: list[TaskInformation]) -> None: ...

    @abstractmethod
    def by_key(self, key: str) -> TaskInformation: ...
