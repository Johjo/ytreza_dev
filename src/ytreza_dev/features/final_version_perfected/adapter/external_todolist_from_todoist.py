from ytreza_dev.features.final_version_perfected.port.external_todolist import ExternalTodolistPort
from ytreza_dev.shared.env_reader import EnvReaderPort
from ytreza_dev.shared.todoist.todoist_api import TodoistAPI


class ExternalTodolistFromTodoist(ExternalTodolistPort):
    def __init__(self, env_reader: EnvReaderPort) -> None:
        self._todoist = TodoistAPI(env_reader.read("TODOIST_API_TOKEN"))

    def close_task(self, url: str, task_id: str) -> None:
        print(f"close task {task_id} on {url}")
        self._todoist.close_task(task_id=task_id)
