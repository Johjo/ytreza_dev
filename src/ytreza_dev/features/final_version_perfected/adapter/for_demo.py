from expression import Nothing
from pyqure import Key

from ytreza_dev.features.final_version_perfected.port.external_todolist import ExternalTodolistPort
from ytreza_dev.features.final_version_perfected.port.task_information_repository import TaskInformationRepositoryPort, \
    TaskInformation
from ytreza_dev.features.final_version_perfected.port.task_reader import TaskFvpReaderPort
from ytreza_dev.features.final_version_perfected.port.task_repository import FvpRepositoryPort
from ytreza_dev.features.final_version_perfected.types import TaskBase, ExternalTask


class TaskInMemory:
    def __init__(self) -> None:
        self._tasks: list[TaskBase] = []

    def all_tasks(self) -> list[TaskBase]:
        return self._tasks

    def save(self, tasks: list[TaskBase]) -> None:
        self._tasks = tasks


class TaskFvpReaderForDemo(TaskFvpReaderPort):
    def __init__(self, memory: TaskInMemory):
        self._memory = memory

    def all_active_tasks(self) -> list[TaskBase]:
        return self._memory.all_tasks()


class FvpRepositoryForDemo(FvpRepositoryPort):
    def all_tasks(self) -> list[TaskBase]:
        return self._memory.all_tasks()

    def __init__(self, memory: TaskInMemory):
        self._memory = memory

    def save(self, tasks: list[TaskBase]) -> None:
        self._memory.save(tasks)


TASK_IN_MEMORY_KEY = Key("task_in_memory", TaskInMemory)


class ExternalTodolistForDemo(ExternalTodolistPort):
    def close_task(self, url: str, task_id: str) -> None:
        pass


class TaskInformationRepositoryForDemo(TaskInformationRepositoryPort):
    def __init__(self):
        self._tasks: dict[str, TaskInformation] = {}

    def by_key(self, key: str) -> TaskInformation:
        return self._tasks[key]

    def save(self, task: list[TaskInformation]) -> None:
        for task in task:
            self._tasks[task.key] = task
