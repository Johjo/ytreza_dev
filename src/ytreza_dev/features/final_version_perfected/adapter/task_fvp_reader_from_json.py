import json
from pathlib import Path

from ytreza_dev.features.final_version_perfected.port.task_reader import TaskFvpReaderPort
from ytreza_dev.features.final_version_perfected.types import TaskBase, TaskNew, TaskNext, TaskLater, TaskNever, Project


class TaskFvpReaderFromJson(TaskFvpReaderPort):
    def __init__(self, json_path: Path) -> None:
        self._json_path = json_path

    def all_active_tasks(self) -> list[TaskBase]:
        tasks = json.loads(self._json_path.read_text(encoding="utf-8"))
        return [self._to_task(task) for task in tasks]

    def _to_task(self, task: dict[str, str]) -> TaskBase:
        match task["status"]:
            case "new":
                return TaskNew(title=task["title"], url=task["url"], id=task["id"])
            case "next":
                return TaskNext(title=task["title"], url=task["url"], id=task["id"])
            case "later":
                return TaskLater(title=task["title"], url=task["url"], id=task["id"])
            case "never":
                return TaskNever(title=task["title"], url=task["url"], id=task["id"])
        raise ValueError(f"Unknown status: {task['status']}")

    @staticmethod
    def _to_project(task) -> Project:
        return Project(key=task["project"]["key"], name=task["project"]["name"])

