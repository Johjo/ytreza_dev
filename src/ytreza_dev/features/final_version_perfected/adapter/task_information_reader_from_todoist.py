from ytreza_dev.features.final_version_perfected.port.task_information_reader import TaskInformationReaderPort, \
    TaskInformation
from ytreza_dev.features.final_version_perfected.types import Project
from ytreza_dev.shared.env_reader import EnvReaderFromEnv
from ytreza_dev.shared.todoist.todoist_api import TodoistAPI, TodoistTask


class TaskInformationReaderFromTodoist(TaskInformationReaderPort):
    def __init__(self, env_reader: EnvReaderFromEnv) -> None:
        self._todoist = TodoistAPI(env_reader.read("TODOIST_API_TOKEN"))

    def by_key(self, key: str) -> TaskInformation:
        task : TodoistTask = self._todoist.task_by_id(task_id=key)
        return TaskInformation(key=task.id, title=task.name, project=Project(key=task.project.id, name=task.project.name), due_date=None)
