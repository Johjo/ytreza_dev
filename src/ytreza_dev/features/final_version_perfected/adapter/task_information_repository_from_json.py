import json
from dataclasses import asdict
from pathlib import Path
from typing import Any

from expression import Nothing

from ytreza_dev.features.final_version_perfected.port.task_information_repository import TaskInformationRepositoryPort, \
    TaskInformation
from ytreza_dev.features.final_version_perfected.types import Project


class TaskInformationRepositoryFromJson(TaskInformationRepositoryPort):
    def __init__(self, json_path: Path) -> None:
        self._json_path = json_path


    def save(self, tasks: list[TaskInformation]) -> None:
        tasks_dict = [self._to_dict(task) for task in tasks]
        self._json_path.write_text(json.dumps(tasks_dict, indent=4))

    @staticmethod
    def _to_dict(task: TaskInformation) -> dict[str, Any]:
        d = asdict(task)
        return d

    def by_key(self, key: str) -> TaskInformation:
        tasks = json.loads(self._json_path.read_text(encoding="utf-8"))
        return [self._to_task_information(task) for task in tasks if task["key"] == key][0]

    @staticmethod
    def _to_task_information(task) -> TaskInformation:
        return TaskInformation(
            key=task["key"],
            title=task["title"],
            project=Project(key=task["project"]["key"], name=task["project"]["name"]),
            url=task["url"],
            due_date=Nothing
        )
