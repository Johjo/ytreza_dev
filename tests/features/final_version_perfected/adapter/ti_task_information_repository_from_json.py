from pathlib import Path

import pytest
from approvaltests import verify
from expression import Nothing

from ytreza_dev.features.final_version_perfected.adapter.task_information_repository_from_json import \
    TaskInformationRepositoryFromJson
from ytreza_dev.features.final_version_perfected.port.task_information_repository import TaskInformation
from ytreza_dev.features.final_version_perfected.types import Project


@pytest.mark.integration
def test_write_in_json() -> None:
    json_path = Path("./data_test/task_repository.json")
    sut = TaskInformationRepositoryFromJson(json_path)

    sut.save([
        TaskInformation(key="1", title="buy the milk", project=Project(key="1", name="Project 1"), due_date=Nothing, url="https://url_1.com"),
        TaskInformation(key="2", title="buy the water", project=Project(key="2", name="Project 2"), due_date=Nothing, url="https://url_2.com"),
    ])

    verify(json_path.read_text(encoding="utf-8"))


@pytest.mark.integration
def test_read_from_json() -> None:
    json_path = Path("./data_test/task_repository.json")
    sut = TaskInformationRepositoryFromJson(json_path)

    expected_task : list[TaskInformation] = [
        TaskInformation(key="1", title="buy the milk", project=Project(key="1", name="Project 1"), due_date=Nothing, url="https://url_1.com"),
        TaskInformation(key="2", title="buy the water", project=Project(key="2", name="Project 2"), due_date=Nothing, url="https://url_2.com"),
    ]
    sut.save(expected_task)

    assert sut.by_key(expected_task[0].key) == expected_task[0]
    assert sut.by_key(expected_task[1].key) == expected_task[1]
