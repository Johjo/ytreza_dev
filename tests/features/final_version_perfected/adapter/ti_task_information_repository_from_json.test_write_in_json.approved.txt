[
    {
        "key": "1",
        "url": "https://url_1.com",
        "title": "buy the milk",
        "project": {
            "key": "1",
            "name": "Project 1"
        },
        "due_date": {
            "tag": "none",
            "none": null
        }
    },
    {
        "key": "2",
        "url": "https://url_2.com",
        "title": "buy the water",
        "project": {
            "key": "2",
            "name": "Project 2"
        },
        "due_date": {
            "tag": "none",
            "none": null
        }
    }
]
